### PR TITLE
add test case for append() bug appending to only last selected elemen…

### PR DESCRIPTION
**Pull request by ianwhitedeveloper**
…t instead of every instance.

The problem was in adjacent() method, a dependency of append().
Root cause: line 31 in adjacent.js: changed fragment.appendChild(n); to fragment.appendChild(n.cloneNode(true)); since appendChild will remove the selected element from the dom and relocate it to the next node in the loop, thereby insuring it will only ever be appended to last of selected nodes.

Edit test case by adding done() as recommended by error thrown by mocha when running test suite.

### CODE REVIEW

**Comment by ianwhitedeveloper** [comment]
@franciscop I also solved this by using `fragment.cloneNode(true) in append.js. Do you have a preference? 

`u.prototype.append = function(html, data) {
  return this.adjacent(html, data, function(node, fragment){
    node.appendChild(fragment.cloneNode(true));
  });
};`

Also, I just realized I incorrectly edited the "Keeps the events when appending" test by added `done()` at the end. It appears events are not being copied using my current solutions. I'll look into it further and I welcome any suggestions.

Thanks.

**Comment by franciscop** [comment]
The `done()` should be inside the `on('click', function(){});` statement if I'm not mistaken, as in:

```js
on('click', function(){ done(); }));
```

Because [the trigger should interrupt the normal flow of the code](http://stackoverflow.com/a/17936225/938236) in a similar way to `setTimeout(function(){}, 0)`. So the `done` should happen inside the event callback and not after the `trigger()`.

**Comment by franciscop** [comment]
So the problem is that the event is not being triggered. BTW I prefer the way in your pull request `n.cloneNode(true)` because it avoids code repetition.

**Comment by ianwhitedeveloper** [comment]
@franciscop I've run into a blocker for this PR. At the moment I can't find a way clone a DOM node's event listeners.

[node.cloneNode(true) only clones the node and it's children, not any event listeners.](http://stackoverflow.com/questions/15408394/how-to-copy-a-dom-node-with-event-listeners)

In chrome there is a `getEventListeners` function that could be used, but no other browser seems to support it. 

According to [this resource](https://coderwall.com/p/b1o2gw/getting-a-list-of-your-eventlisteners-so-you-can-trigger-it-programmatically) this is how jQuery solves this problem:

> jQuery, YUI, Moo etc. keeps a copy of the eventListeners when you add them to elements so that you have access to a list.
....
> According to DOM 3 Events, when you used addEventListener there should be an event listener list (EventListenerList DOM 3 Events), but it has yet to be implemented by any browser (Sept, 2012). Until then there is no easy getEventListener method. If you want to programmatically triggering events, keep a copy of a list globally or use a library.

**Comment by franciscop** [comment]
I've been searching and it's true, it seems to be a really big issue. The HTML elements have a property called `_e` where I'm storing the events, so if they are copied we could just transverse all the elements and reasign them

**Comment by ianwhitedeveloper** [comment]
Ah I see. I'm a dummy I completely missed that in the source code. Thanks @franciscop!

File: src/plugins/adjacent/adjacent.js [hunk 1]
```diff
@@ -28,7 +28,19 @@ u.prototype.adjacent = function(html, data, callback) {
 
       return u(part).nodes;
     }).each(function(n){
-      fragment.appendChild(n);
+      var cloneWithEvents = n.cloneNode(true),
```
Comment (ianwhitedeveloper) [inline src/plugins/adjacent/adjacent.js]: TODO: implement u.prototype.clone method

Comment (franciscop) [inline src/plugins/adjacent/adjacent.js]: I don't have much time now, but would it be possible to expand it with this?

```js
u(n).find('*').filter(function(node){ return node._e; }).each(function(){
  for ...
});
```

That way we also copy the children of `n` events, not only n's events. But it's getting really ugly really fast, so I totally agree with adding a `.clone()` method.

Comment (ianwhitedeveloper) [inline src/plugins/adjacent/adjacent.js]: Yeah no problem and thank you for the input that will speed up my solution. Should I be withdrawing pr's while they're still in progress?

Comment (franciscop) [inline src/plugins/adjacent/adjacent.js]: No way, I love seeing the process real-time-ish

**Comment by ianwhitedeveloper** [comment]
@franciscop  I believe I've fixed this issue. The fix relies on the clone method created in this PR https://github.com/umbrellajs/umbrella/pull/66

Once merged I will update this PR with the fix.  Cheers.

**Pull request merged**